### DISCUSSION TITLE
persistent session, QOS 1 and failover

### DISCUSSION THREAD

**Discussion by sushiMix** [emqx/emqx#11471] [Q&A] [answered]
Hello I'm doing some tests with EMQX 5.X to test the guaranteed delivery.

I'm performing some resilience test to ensure not losing messages in QOS 1 .

My use case is :
I deploy 1 static cluster with 3 nodes (as seed/core nodes).

I add 3 producer which connect to a different node sending incremental data in QOS 1 (on topic test/load//serverX where 1<=X<=3).
`
i=1
srv=serverX
while [ "$i" -ne 1000000 ]
do
    m=$(printf '{ "data" : %d }' "$i")
    while
      pub -h "${srv}" -q 1 -p 1883 -V 5 -D publish message-expiry-interval 600 -t "test/load/$srv" -m "$m"
      [ $? -ne 0 ]
    do :; done
    i=$((i + 1))
done
`
I add Y consumers listening to the 3 topics (subscribe with qos 1 and clean session to false with a consistent clientId) and checking that the message are received without loss (comparing values between messages where delta (new - old) should be 1).
Listeners are web-sockets ones with mqttjs library.

When doing some network loss tests between client and servers I see no errors (the reconnection is well done and I have no message "loss" ).

Now I'm trying to kill an EMQX node and I notice that some messages are lost in such use case (when I'm killiing the node where the subscriber is) Killing a node where only producers are connected does not seem an issue.

Does somebody already noticed it ? Is there some recommendation about tuning ?

**Comment by lafirest** [comment]
@sushiMix 
The received messages are stored in memory, so if you kill the node, it will lose. The persistent session is in development.

**Reply by sushiMix** [reply]
Do you mean, the cluster deployment does not forward the message to other nodes to avoid loss of message when QOS > 0.

**Comment by ieQu1** [comment, marked as answer]
> Now I'm trying to kill an EMQX node and I notice that some messages are lost in such use case (when I'm killiing the node where the subscriber is) Killing a node where only producers are connected does not seem an issue.

> Does somebody already noticed it ? Is there some recommendation about tuning ?

Currently EMQX creates a proxy process that takes place of the consumer when it's down. This process is only present on one EMQX node. So if you kill this node, the process is gone. We're working on a durable message storage that is free of this issue.
